### PR TITLE
make types within `generate_solution_type` macro explicit

### LINKED ISSUE #8444: Rename `generate_solution_type!` such that it shows exactly the types
This is an example of a solution type generation: 

```
sp_npos_elections::generate_solution_type!(
	#[compact]
	pub struct NposSolution16::<u32, u16, sp_runtime::PerU16>(16)
	// -------------------- ^^ <NominatorIndex, ValidatorIndex, Accuracy>
);
```

The fact that I have to document it, and double check it is exactly showing the issue here. the syntax of the macro should be such that it clearly shows these. After all, the entire `::` and `<_>` is fake and I am manually parsing it. Maybe something like 

```
sp_npos_elections::generate_solution_type!(
	#[compact]
	pub struct NposSolution16::<Voter = u32, Target = u16, Accuracy = sp_runtime::PerU16>(16)
);
```

### PR DESCRIPTION
**Pull request by coriolinus**
Closes #8444.

polkadot companion: https://github.com/paritytech/polkadot/pull/2707

Just changes the parsing logic for that macro; does not change any
emitted code. The associated types associated with the macro now
require explicit, keyword-style declaration.

**Note**: while the new syntax looks like keyword arguments from a language such as Python,
unlike real keyword arguments, the order still matters. Checks in the syntax parser ensure that 
the right arguments appear in the right order. 

**Old**:

```rust
sp_npos_elections::generate_solution_type!(
	#[compact]
	pub struct TestCompact::<VoterIndex, TargetIndex, PerU16>(16)
);
```

**New**:

```rust
sp_npos_elections::generate_solution_type!(
	#[compact]
	pub struct TestCompact::<VoterIndex = VoterIndex, CandidateIndex = TargetIndex, Accuracy = PerU16>(16)
);
```

### CODE REVIEW

**Review by kianenigma — APPROVED** [review]
In the jargon of `npos_election` stuff, we have `voters` and `targets`, let's stick to that and rename `CandidateIndex` to `TargetIndex`, otherwise looks good, thanks!

**Comment by kianenigma** [comment]
>Note: while the new syntax looks like keyword arguments from a language such as Python,
unlike real keyword arguments, the order still matters. Checks in the syntax parser ensure that
the right arguments appear in the right order.

Would be good to have a test to make sure 

```
sp_npos_elections::generate_solution_type!(
	#[compact]
	pub struct TestCompact::<TargetIndex = X, VoterIndex = Y, Accuracy = PerU16>(16)
);
```

fails.

File: primitives/npos-elections/compact/src/lib.rs [hunk 1]
```diff
@@ -66,10 +71,11 @@ pub(crate) fn syn_err(message: &'static str) -> syn::Error {
 /// for numbers will be used, similar to how `parity-scale-codec`'s `Compact` works.
 ///
 /// ```ignore
```
Comment (thiolliere) [inline primitives/npos-elections/compact/src/lib.rs:73]: sidenote: why is this test ignored ?

**Review by thiolliere — APPROVED** [review]
looks good to me

File: primitives/npos-elections/compact/src/lib.rs (hunk 1 above)
Comment (kianenigma) [reply primitives/npos-elections/compact/src/lib.rs:73]: not sure, can we bring it back @coriolinus?

Comment (coriolinus) [reply primitives/npos-elections/compact/src/lib.rs]: Sure: a2d3ac381

Comment (kianenigma) [reply primitives/npos-elections/compact/src/lib.rs]: sweet

**Comment by coriolinus** [comment]
bot merge

**Comment by coriolinus** [comment]
bot merge

**Comment by coriolinus** [comment]
Looking for recommendations for how to proceed with this. The polkadot companion stage fails because that PR is not mergeable. The only reason I can see that https://github.com/paritytech/polkadot/pull/2707/ is not mergeable is that its `gitlab-test-linux-stable` CI check fails. 

I can replicate the failure in that CI check only if I do not have the `~/.cargo/config` `paths` set appropriately for a companion. When it's set as required, then everything builds as expected locally.

**Comment by thiolliere** [comment]
did you merge polkadot master into the companion ?

**Comment by bkchr** [comment]
The companion wasn't approved.

**Comment by kianenigma** [comment]
> I can replicate the failure in that CI check only if I do not have the ~/.cargo/config paths set appropriately for a companion.

The Way that polkadot companions are merged is that their own CI is supposed to fail, so what you are seeing is correct. For a companion, you check the build status in the substrate `polkadot-build` CI job. 

Once the substrate PR is merged, the bot does a `cargo update -p sp-io` in polkadot, then run again.

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**